pkg/token/serviceprincipaltoken_test.go

```diff
@@ -0,0 +1,232 @@
+	return rec, rec.GetDefaultClient()
+}
+
+func customMatcher(r *http.Request, i cassette.Request) bool {
+	id := os.Getenv(tenantID)
+	if id == "" {
+		id = "00000000-0000-0000-0000-000000000000"
+	}
+	switch os.Getenv(vcrMode) {
+	case vcrModeRecordOnly:
+	default:
+		r.URL.Path = strings.ReplaceAll(r.URL.Path, id, tenantID)
+	}
+	return cassette.DefaultMatcher(r, i)
+}
+
+// Get go-vcr record mode from enviroment variable
+func getVCRMode() recorder.Mode {
+	switch os.Getenv(vcrMode) {
+	case vcrModeRecordOnly:
+		return recorder.ModeRecordOnly
+	default:
+		return recorder.ModeReplayOnly
+	}
+}
+
+func TestServicePrincipalLoginVCR(t *testing.T) {
+	testCase := []struct {
+		cassetteName  string
+		description   string
+		p             *servicePrincipalToken
+		expectedError error
+	}{
+		{
+			cassetteName: "BadSecretVCR",
+			description:  "Test using incorrect secret value",
+			p: &servicePrincipalToken{
+				clientID:     os.Getenv(clientID),
+				clientSecret: badSecret,
+				resourceID:   os.Getenv(resourceID),
+				tenantID:     os.Getenv(tenantID),
+			},
+			expectedError: fmt.Errorf("ClientSecretCredential authentication failed"),
+		},
+		{
+			cassetteName: "SecretTokenVCR",
+			description:  "Test using service principal secret value to get token",
+			p: &servicePrincipalToken{
+				clientID:     os.Getenv(clientID),
+				clientSecret: os.Getenv(clientSecret),
+				resourceID:   os.Getenv(resourceID),
+				tenantID:     os.Getenv(tenantID),
+			},
+			expectedError: nil,
+		},
+		{
+			cassetteName: "CertTokenVCR",
+			description:  "Test ",
+			p: &servicePrincipalToken{
+				clientID:           os.Getenv(clientID),
+				clientCert:         os.Getenv(clientCert),
+				clientCertPassword: os.Getenv(clientCertPass),
+				resourceID:         os.Getenv(resourceID),
+				tenantID:           os.Getenv(tenantID),
+			},
+			expectedError: nil,
+		},
+	}
+
+	for _, tc := range testCase {
+		t.Run(tc.description, func(t *testing.T) {
+			vcrRecorder, httpClient := getVCRHttpClient(fmt.Sprintf("testdata/%s", tc.cassetteName))
+
+			clientOpts := azcore.ClientOptions{
+				Cloud:     cloud.AzurePublic,
+				Transport: httpClient,
+			}
+			if tc.p.clientID == "" {
+				tc.p.clientID = clientID
+			}
+			if tc.p.clientSecret == "" && strings.Contains(tc.cassetteName, "Secret") {
+				tc.p.clientSecret = clientSecret
+			}
+			if tc.p.clientCert == "" {
+				tc.p.clientCert = "testdata/testCert.pfx"
+			}
+			if tc.p.clientCertPassword == "" {
+				tc.p.clientCertPassword = "TestPassword"
+			}
+			if tc.p.resourceID == "" {
+				tc.p.resourceID = resourceID
+			}
+			if tc.p.tenantID == "" {
+				tc.p.tenantID = "00000000-0000-0000-0000-000000000000"
+			}
+
+			token, err := tc.p.TokenOptions(&clientOpts)
+			defer vcrRecorder.Stop()
+			if err != nil {
+				if !ErrorContains(err, tc.expectedError.Error()) {
+					t.Errorf("expected error %v, but got %v", tc.expectedError.Error(), err)
+				}
+			} else {
+				if token.AccessToken == "" {
```

Comment:
i think we should modify the access token during `BeforeResponseReplayHook` and verify that in the unit test

Reply:
Added by generating a new UUID on each test and verifying it is returned.